### PR TITLE
R2 final review

### PR DESCRIPTION
**Pull request by GoodluckH**

### CODE REVIEW

**Comment by ServiceDog** [comment]
## Documents
- [ ] ICD
  - [ ] Current interfaces (grpc & rest)
  - [ ] Current relevant document links
- [ ] SDD
  - [ ] Description of grpc & rest handler implementations
  - [ ] Current sequence diagrams where applicable
  - [ ] Current relevant document links
- [ ] CONOPS
  - [ ] Current information
  - [ ] Bibliography (if any citations)
- [ ] README (see svc-cargo for example)
  - [ ] Badges
  - [ ] Accurate installation instructions
  - [ ] Current links to other documents
  - [ ] Arrow DAO links
- [ ] Any instance of :construction: should genuinely be under construction

## Code (not applicable to examples/ folder)
- [ ] No unhandled unwrap()s
- [ ] Accurate (!) code comments on all functions
- [ ] No `# Arguments` and `# Returns` in function comments
- [ ] Function arguments checked where possible
- [ ] Sufficient logger (info, warn, and debug) messages
- [ ] Sensible Types (unsigned vs. signed)
- [ ] No `#[allow(dead_code)]` without good reason
- [ ] End all error, info, and debug messages with punctuation (periods, etc.)
- [ ] No FIXMEs or TODOs without a projected release (`TODO R4: etc.`)

## REST API
- [ ] Return codes accurately documented with openapi
- [ ] Return codes make sense
- [ ] Sufficiently descriptive error messages
- [ ] Do NOT reflect a user's bad input back at them in replies.
    - i.e. "Warning: '{}' is not a valid entry." should instead be "Warning: invalid entry."
    - Use debug messages to see the invalid argument.

## Cargo.toml(s)
- [ ] Package versions correct
- [ ] No minor versions specified
- [ ] Links to crates.io package, not git branch
- [ ] Must have `license-file = "LICENSE"` under `[package]`

File: server/src/main.rs [hunk 1]
```diff
@@ -26,22 +40,28 @@ pub struct ArrowPricing {}
...
-        grpc_info!("Received request: {:?}", request);
         let query = request.into_inner();
-        grpc_debug!("Getting pricing for query: {:?}", query);
-        let price = get_pricing(query);
-        grpc_debug!("Pricing computed: {}", price);
-        let response = pricing_grpc::PricingResponse { price };
-        grpc_debug!("Returning response: {:?}", response);
-        grpc_info!("Successfully computed pricing; returning response");
-        Ok(tonic::Response::new(response))
+        info!("Received pricing request");
```
Comment (ServiceDog) [inline server/src/main.rs:49]: make this a debug call, and end it with punctuation

also nice to prepend (get_pricing) to the message so it's easily grep'd in logs

File: server/src/engine.rs [hunk 2]
```diff
@@ -1,31 +1,100 @@
...
+/// allow for a trip that involves different types of services.
+///
+/// This may change in the future as per business requirements.
+///
+/// # Returns
+/// If no errors occur, returns a vector of prices in dollars,
+/// corresponding to the order of the input requests.
+pub fn get_pricing(query: PricingRequests) -> Result<Vec<f32>, PricingError> {
+    let requests = query.requests;
+    info!("Getting pricing for {:?} requests", requests.len());
```
Comment (ServiceDog) [inline server/src/engine.rs:50]: consider making this a debug call with (get_pricing) prepended

File: server/src/engine.rs [hunk 3]
```diff
@@ -1,31 +1,100 @@
...
-            0.0
+            debug!("Error parsing service type");
+            Err(PricingError::UnknownServiceType)
+        }
+    }
+}
+
+/// Check that all pricing requests have the same service type.
+fn check_pricing_requests(requests: &[PricingRequest]) -> Result<(), PricingError> {
+    if requests.is_empty() {
```
Comment (ServiceDog) [inline server/src/engine.rs:84]: add a debug statement with "(check_pricing_requests) entry"

**Review by ServiceDog — CHANGES_REQUESTED** [review]

File: server/src/engine.rs (hunk 3 above)
Comment (GoodluckH) [reply server/src/engine.rs:92]: [17d9bf3](https://github.com/Arrow-air/svc-pricing/pull/19/commits/17d9bf3b229051e65e10107c1e7d5756c5de34e0)

File: server/src/engine.rs (hunk 2 above)
Comment (GoodluckH) [reply server/src/engine.rs]: [17d9bf3](https://github.com/Arrow-air/svc-pricing/pull/19/commits/17d9bf3b229051e65e10107c1e7d5756c5de34e0)

File: server/src/main.rs (hunk 1 above)
Comment (GoodluckH) [reply server/src/main.rs:49]: [17d9bf3](https://github.com/Arrow-air/svc-pricing/pull/19/commits/17d9bf3b229051e65e10107c1e7d5756c5de34e0)

File: docs/icd.md [hunk 4]
```diff
@@ -26,6 +26,7 @@ Document | Description
 [High-Level Services ICD](https://github.com/Arrow-air/se-services/blob/develop/docs/icd.md) | Arrow Services team's top-level ICD.
 [Concept of Operations - `svc-pricing`](conops.md) | Concept of Operations for `svc-pricing`.
 [Software Design Document - `svc-pricing`](./sdd.md) | Software Design Document for `svc-pricing`.
+[Requirements - `svc-pricing`](https://nocodb.arrowair.com/dashboard/#/nc/view/045288a8-3875-4429-bdaa-9f578275adef) | Requirements for `svc-pricing`.
 
 ## Frameworks
```
[the remark this answers is not in the document]
Comment (GoodluckH) [reply docs/icd.md:31]: [17d9bf3](https://github.com/Arrow-air/svc-pricing/pull/19/commits/17d9bf3b229051e65e10107c1e7d5756c5de34e0)

File: server/src/engine.rs [hunk 5]
```diff
@@ -1,31 +1,100 @@
-//! Implementation of pricing logics based on the ride type.
+//! Implementation of pricing logics based on the service type.
```
[the remark this answers is not in the document]
Comment (GoodluckH) [reply server/src/engine.rs:1]: [17d9bf3](https://github.com/Arrow-air/svc-pricing/pull/19/commits/17d9bf3b229051e65e10107c1e7d5756c5de34e0)

**Review by ServiceDog — APPROVED** [review]

**Pull request merged**